std/range/package.d

```diff
@@ -1364,183 +1364,189 @@ Params:
 
 Returns:
     A range type dependent on $(D R1) and $(D R2).
-
-Bugs:
-    $(BUGZILLA 14660)
  */
 auto choose(R1, R2)(bool condition, R1 r1, R2 r2)
 if (isInputRange!(Unqual!R1) && isInputRange!(Unqual!R2) &&
     !is(CommonType!(ElementType!(Unqual!R1), ElementType!(Unqual!R2)) == void))
 {
-    static struct Result
+    return ChooseResult!(R1, R2)(condition, r1, r2);
+}
+
+private struct ChooseResult(R1, R2)
+{
+    import std.traits : hasElaborateCopyConstructor, hasElaborateDestructor;
+
+    private union
     {
-        import std.algorithm.comparison : max;
-        import std.algorithm.internal : addressOf;
-        import std.traits : hasElaborateCopyConstructor, hasElaborateDestructor;
+        R1 _r1;
+        R2 _r2;
+    }
+    private bool r1Chosen;
 
-        private union
-        {
-            void[max(R1.sizeof, R2.sizeof)] buffer = void;
-            void* forAlignmentOnly = void;
-        }
-        private bool condition;
-        private @property ref R1 r1()
+    private static auto ref actOnChosen(alias foo, ExtraArgs ...)(ref ChooseResult r,
+            auto ref ExtraArgs extraArgs)
+    {
+        if (r.r1Chosen)
+            return foo(r._r1, extraArgs);
+        else
+            return foo(r._r2, extraArgs);
+    }
+
+    this(bool r1Chosen, R1 r1, R2 r2)
+    {
+        import std.conv : emplace;
+
+        // This should be the only place r1Chosen is ever assigned
+        // independently
+        this.r1Chosen = r1Chosen;
+        if (r1Chosen)
         {
-            assert(condition);
-            return *cast(R1*) buffer.ptr;
+            _r2 = R2.init;
```

Comment:
remove

Reply:
dmd doesn't like that. `Error: field r2 must be initialized in constructor, because it is nested struct`